<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ mcATAC</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">mcATAC</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/mcATAC.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/analysis.html">peaks</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/tanaylab/mcATAC/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="ATAC-class.html">ATAC-class</a></code> 
        </dt>
        <dd>ATAC objects</dd>
      </dl><dl><dt>
          
          <code><a href="ATACPeaks-class.html">ATACPeaks-class</a></code> 
        </dt>
        <dd>ATACPeaks objects</dd>
      </dl><dl><dt>
          
          <code><a href="ATACPeaks.html">McPeaks-class</a></code> 
        </dt>
        <dd>McPeaks</dd>
      </dl><dl><dt>
          
          <code><a href="ATACTracks-class.html">ATACTracks-class</a></code> 
        </dt>
        <dd>ATACTracks objects</dd>
      </dl><dl><dt>
          
          <code><a href="ATACTracks.html">McTracks-class</a></code> 
        </dt>
        <dd>McTracks</dd>
      </dl><dl><dt>
          
          <code><a href="PeakIntervals.html">PeakIntervals()</a></code> 
        </dt>
        <dd>Construct a PeakIntervals object from misha intervals</dd>
      </dl><dl><dt>
          
          <code><a href="ScCounts-class.html">ScCounts-class</a></code> 
        </dt>
        <dd>ScCounts object</dd>
      </dl><dl><dt>
          
          <code><a href="ScCounts.html">McCounts-class</a></code> 
        </dt>
        <dd>McCounts object</dd>
      </dl><dl><dt>
          
          <code><a href="add_mc_metadata.html">add_mc_metadata()</a></code> 
        </dt>
        <dd>Add per-metacell metadata to an McPeaks object</dd>
      </dl><dl><dt>
          
          <code><a href="add_mc_rna.html">add_mc_rna()</a></code> 
        </dt>
        <dd>Add per-metacell gene expression data to an McPeaks object</dd>
      </dl><dl><dt>
          
          <code><a href="annotate_intervals.html">annotate_intervals()</a></code> 
        </dt>
        <dd>Annotate an intervals set</dd>
      </dl><dl><dt>
          
          <code><a href="annotate_peaks.html">annotate_peaks()</a></code> 
        </dt>
        <dd>Annotate ATAC peaks</dd>
      </dl><dl><dt>
          
          <code><a href="atac_ignore_cells.html">atac_ignore_cells()</a></code> 
        </dt>
        <dd>Remove cells</dd>
      </dl><dl><dt>
          
          <code><a href="atac_ignore_peaks.html">atac_ignore_peaks()</a></code> 
        </dt>
        <dd>Set ignored (i.e. blacklisted) peaks</dd>
      </dl><dl><dt>
          
          <code><a href="bam_file_chromosomes.html">bam_file_chromosomes()</a></code> 
        </dt>
        <dd>Get chromosomes from a bam file</dd>
      </dl><dl><dt>
          
          <code><a href="calc_prom_rna_cor.html">calc_prom_rna_cor()</a></code> 
        </dt>
        <dd>Calculate the cross-correlation between promoter accessibility and RNA expression</dd>
      </dl><dl><dt>
          
          <code><a href="calculate_d_stats.html">calculate_d_stats()</a></code> 
        </dt>
        <dd>Calculate Kolmogorov-Smirnov D statistics between two interval sets with motif energies</dd>
      </dl><dl><dt>
          
          <code><a href="call_peaks.html">call_peaks()</a></code> 
        </dt>
        <dd>Call peaks from ATAC marginals track</dd>
      </dl><dl><dt>
          
          <code><a href="check_dependencies.html">check_dependencies()</a></code> 
        </dt>
        <dd>Test if mcATAC has the required dependencies</dd>
      </dl><dl><dt>
          
          <code><a href="clear_cache.html">clear_cache()</a></code> 
        </dt>
        <dd>Clear the ATAC track matrix cache</dd>
      </dl><dl><dt>
          
          <code><a href="create_bundle.html">create_bundle()</a></code> 
        </dt>
        <dd>Create a bundle for the shiny app in order to run it with shiny server</dd>
      </dl><dl><dt>
          
          <code><a href="download_pbmc_example_data.html">download_pbmc_example_data()</a></code> 
        </dt>
        <dd>Download example data from 10x</dd>
      </dl><dl><dt>
          
          <code><a href="export_atac_clust_misha.html">export_atac_clust_misha()</a></code> 
        </dt>
        <dd>Generate a misha track for each atac metacell cluster</dd>
      </dl><dl><dt>
          
          <code><a href="export_atac_clust_ucsc.html">export_atac_clust_ucsc()</a></code> 
        </dt>
        <dd>Generate a ucsc genome browser file for each metacell cluster</dd>
      </dl><dl><dt>
          
          <code><a href="export_to_h5ad.html">export_to_h5ad()</a></code> 
        </dt>
        <dd>Write a McPeaks or ScPeaks object to an h5ad file</dd>
      </dl><dl><dt>
          
          <code><a href="filter_features.html">filter_features()</a></code> 
        </dt>
        <dd>Filter features by summary statistics</dd>
      </dl><dl><dt>
          
          <code><a href="find_blacklist_overlaps.html">find_blacklist_overlaps()</a></code> 
        </dt>
        <dd>Find overlaps with ENCODE blacklists</dd>
      </dl><dl><dt>
          
          <code><a href="gen_atac_mc_clust.html">gen_atac_mc_clust()</a></code> 
        </dt>
        <dd>Cluster metacells based on atac profiles using the k-means algorithm</dd>
      </dl><dl><dt>
          
          <code><a href="gen_atac_peak_clust.html">gen_atac_peak_clust()</a></code> 
        </dt>
        <dd>Cluster atac peaks based on atac distributions</dd>
      </dl><dl><dt>
          
          <code><a href="gen_promoter_features.html">gen_promoter_features()</a></code> 
        </dt>
        <dd>Select only peaks that are close to a gene promoter</dd>
      </dl><dl><dt>
          
          <code><a href="gen_random_genome_peak_motif_matrix.html">gen_random_genome_peak_motif_matrix()</a></code> 
        </dt>
        <dd>Generate random genome motif PSSM matrix</dd>
      </dl><dl><dt>
          
          <code><a href="generate_motif_pssm_matrix.html">generate_motif_pssm_matrix()</a></code> 
        </dt>
        <dd>Generate peak-motif matrix</dd>
      </dl><dl><dt>
          
          <code><a href="generate_pheatmap_annotation.html">generate_pheatmap_annotation()</a></code> 
        </dt>
        <dd>Function to generate generic pheatmap annotation</dd>
      </dl><dl><dt>
          
          <code><a href="get_available_pssms.html">get_available_pssms()</a></code> 
        </dt>
        <dd>Function for validating <code>pssm_path</code> and getting all/selected PSSM datasets and tracks from path/gdb</dd>
      </dl><dl><dt>
          
          <code><a href="get_gene_promoter.html">get_gene_promoter()</a></code> 
        </dt>
        <dd>Get an intervals set of a gene promoter.</dd>
      </dl><dl><dt>
          
          <code><a href="get_genes_atac_fp.html">get_genes_atac_fp()</a></code> 
        </dt>
        <dd>Return a relative fold change matrix of ATAC peaks for a list of genes</dd>
      </dl><dl><dt>
          
          <code><a href="get_peak_coverage_stats.html">get_peak_coverage_stats()</a></code> 
        </dt>
        <dd>Calculate coverage statistics of the peaks</dd>
      </dl><dl><dt>
          
          <code><a href="get_promoter_peaks.html">get_promoter_peaks()</a></code> 
        </dt>
        <dd>Get the names of the peaks close to a gene promoter</dd>
      </dl><dl><dt>
          
          <code><a href="get_quantile_cov_thresh.html">get_quantile_cov_thresh()</a></code> 
        </dt>
        <dd>Calculate the threshold for calling peaks</dd>
      </dl><dl><dt>
          
          <code><a href="get_rna_egc.html">get_rna_egc()</a></code> 
        </dt>
        <dd>Get the normalized RNA expression matrix (egc) from a McPeaks object</dd>
      </dl><dl><dt>
          
          <code><a href="get_rna_fp.html">get_rna_fp()</a></code> 
        </dt>
        <dd>Get enrichment of normalized counts of gene expression over the median</dd>
      </dl><dl><dt>
          
          <code><a href="get_rna_marker_matrix.html">get_rna_marker_matrix()</a></code> 
        </dt>
        <dd>Get enrichment matrix for marker genes</dd>
      </dl><dl><dt>
          
          <code><a href="get_rna_markers.html">get_rna_markers()</a></code> 
        </dt>
        <dd>Calculate gene markers of metacells gene expression</dd>
      </dl><dl><dt>
          
          <code><a href="get_track_total_coverage.html">get_track_total_coverage()</a></code> 
        </dt>
        <dd>Return the total coverage of a track</dd>
      </dl><dl><dt>
          
          <code><a href="gintervals.extend_left.html">gintervals.extend_left()</a></code> 
        </dt>
        <dd>Extend the coordinates of an intervals set to the left.</dd>
      </dl><dl><dt>
          
          <code><a href="gintervals.extend_right.html">gintervals.extend_right()</a></code> 
        </dt>
        <dd>Extend the coordinates of an intervals set to the right.</dd>
      </dl><dl><dt>
          
          <code><a href="gintervals.shift_left.html">gintervals.shift_left()</a></code> 
        </dt>
        <dd>Shift the coordinates of an intervals set to the left.</dd>
      </dl><dl><dt>
          
          <code><a href="gintervals.shift_right.html">gintervals.shift_right()</a></code> 
        </dt>
        <dd>Shift the coordinates of an intervals set to the right.</dd>
      </dl><dl><dt>
          
          <code><a href="gintervals.zoom_in.html">gintervals.zoom_in()</a></code> 
        </dt>
        <dd>Zoom-in for an intervals set.</dd>
      </dl><dl><dt>
          
          <code><a href="gintervals.zoom_out.html">gintervals.zoom_out()</a></code> 
        </dt>
        <dd>Zoom-out for an intervals set.</dd>
      </dl><dl><dt>
          
          <code><a href="has_cell_type.html">has_cell_type()</a></code> 
        </dt>
        <dd>Does the McPeaks object contain per-metacell cell type annotation</dd>
      </dl><dl><dt>
          
          <code><a href="has_cell_type_colors.html">has_cell_type_colors()</a></code> 
        </dt>
        <dd>Does the McPeaks object contain cell type color annotation</dd>
      </dl><dl><dt>
          
          <code><a href="has_rna.html">has_rna()</a></code> 
        </dt>
        <dd>Does the McPeaks object contain per-metacell gene expression data?</dd>
      </dl><dl><dt>
          
          <code><a href="identify_dynamic_peaks.html">identify_dynamic_peaks()</a></code> 
        </dt>
        <dd>Find dynamic peaks in McPeaks matrix</dd>
      </dl><dl><dt>
          
          <code><a href="import_atac_marginal.html">import_atac_marginal()</a></code> 
        </dt>
        <dd>Import marginal of ATAC counts to a misha track</dd>
      </dl><dl><dt>
          
          <code><a href="import_from_10x.html">import_from_10x()</a></code> 
        </dt>
        <dd>Create an ScPeaks object from 10x directory</dd>
      </dl><dl><dt>
          
          <code><a href="import_from_h5ad.html">import_from_h5ad()</a></code> 
        </dt>
        <dd>Create an ScPeaks,McPeaks object from an h5ad file</dd>
      </dl><dl><dt>
          
          <code><a href="import_from_matrix.html">import_from_matrix()</a></code> 
        </dt>
        <dd>Create an ScPeaks,McPeaks object from matrix and peaks</dd>
      </dl><dl><dt>
          
          <code><a href="infer_track_order.html">infer_track_order()</a></code> 
        </dt>
        <dd>Infer order of metacell track names by metacell number</dd>
      </dl><dl><dt>
          
          <code><a href="initialize-McPeaks-method.html">initialize(<i>&lt;McPeaks&gt;</i>)</a></code> 
        </dt>
        <dd>Construct a new McPeaks object</dd>
      </dl><dl><dt>
          
          <code><a href="initialize-ScPeaks-method.html">initialize(<i>&lt;ScPeaks&gt;</i>)</a></code> 
        </dt>
        <dd>Construct a new ScPeaks object</dd>
      </dl><dl><dt>
          
          <code><a href="mc_hclust_rna.html">mc_hclust_rna()</a></code> 
        </dt>
        <dd>Hierarchical clustering of metacells using RNA markers</dd>
      </dl><dl><dt>
          
          <code><a href="mc_order.html">mc_order()</a></code> 
        </dt>
        <dd>Order the metacells in an MC object</dd>
      </dl><dl><dt>
          
          <code><a href="mc_order_by_rna.html">mc_order_by_rna()</a></code> 
        </dt>
        <dd>Order metacells by RNA marker clustering</dd>
      </dl><dl><dt>
          
          <code><a href="mcc_extract_to_df.html">mcc_extract_to_df()</a></code> 
        </dt>
        <dd>Extract McCounts to a tidy data frame</dd>
      </dl><dl><dt>
          
          <code><a href="mcc_marginal.html">mcc_marginal()</a></code> 
        </dt>
        <dd>Return the total coverage of each non-zero coordinate in an McCounts object</dd>
      </dl><dl><dt>
          
          <code><a href="mcc_metacell_total_cov.html">mcc_metacell_total_cov()</a></code> 
        </dt>
        <dd>Return the total coverage of each metacell in an McCounts object</dd>
      </dl><dl><dt>
          
          <code><a href="mcc_normalize_metacells.html">mcc_normalize_metacells()</a></code> 
        </dt>
        <dd>Normalize each metacell in a McCounts object by its total counts</dd>
      </dl><dl><dt>
          
          <code><a href="mcc_read.html">mcc_read()</a></code> 
        </dt>
        <dd>Read an McCounts object from a directory</dd>
      </dl><dl><dt>
          
          <code><a href="mcc_to_marginal_track.html">mcc_to_marginal_track()</a></code> 
        </dt>
        <dd>Create a track with smoothed marginal counts from a McCounts object</dd>
      </dl><dl><dt>
          
          <code><a href="mcc_to_mcatac.html">mcc_to_mcatac()</a></code> 
        </dt>
        <dd>Create an McPeaks object from an McCounts object</dd>
      </dl><dl><dt>
          
          <code><a href="mcc_to_tracks.html">mcc_to_tracks()</a></code> 
        </dt>
        <dd>Create a misha track for each metacell in a McCounts object</dd>
      </dl><dl><dt>
          
          <code><a href="mct_create.html">mct_create()</a></code> 
        </dt>
        <dd>Create an McTracks object</dd>
      </dl><dl><dt>
          
          <code><a href="mct_diff_access_on_hc.html">mct_diff_access_on_hc()</a></code> 
        </dt>
        <dd>Return a mask matrix with differential cluster accessibility (DCA)</dd>
      </dl><dl><dt>
          
          <code><a href="mct_get_mat.html">mct_get_mat()</a></code> 
        </dt>
        <dd>Get an ATAC matrix from a McTracks object</dd>
      </dl><dl><dt>
          
          <code><a href="mct_ls_cached_regions.html">mct_ls_cached_regions()</a></code> 
        </dt>
        <dd>List the cached regions of an McTracks object</dd>
      </dl><dl><dt>
          
          <code><a href="mct_plot_region.html">mct_plot_region()</a></code> 
        </dt>
        <dd>Plot a genomic region</dd>
      </dl><dl><dt>
          
          <code><a href="name_enhancers.html">name_enhancers()</a></code> 
        </dt>
        <dd>Name ATAC peaks from TAD names</dd>
      </dl><dl><dt>
          
          <code><a href="order_tracks.html">order_tracks()</a></code> 
        </dt>
        <dd>Order metacell tracks by metacell number</dd>
      </dl><dl><dt>
          
          <code><a href="peak_names.html">peak_names()</a></code> 
        </dt>
        <dd>Return the names of PeakIntervals</dd>
      </dl><dl><dt>
          
          <code><a href="plot_atac_atac_cor.html">plot_atac_atac_cor()</a></code> 
        </dt>
        <dd>Plot a correlation matrix of ATAC metacells</dd>
      </dl><dl><dt>
          
          <code><a href="plot_atac_peak_map.html">plot_atac_peak_map()</a></code> 
        </dt>
        <dd>Plot normalized accessibility of peaks over metacells, ordered by clustering</dd>
      </dl><dl><dt>
          
          <code><a href="plot_atac_rna.html">plot_atac_rna()</a></code> 
        </dt>
        <dd>Plot a scatter of gene expression vs an atac profile</dd>
      </dl><dl><dt>
          
          <code><a href="plot_atac_rna_cor.html">plot_atac_rna_cor()</a></code> 
        </dt>
        <dd>Plot a cross-correlation matrix between RNA metacells and ATAC metacell scores</dd>
      </dl><dl><dt>
          
          <code><a href="plot_atac_rna_markers.html">plot_atac_rna_markers()</a></code> 
        </dt>
        <dd>Plot RNA and ATAC marker matrix side by side</dd>
      </dl><dl><dt>
          
          <code><a href="plot_marginal_coverage.html">plot_marginal_coverage()</a></code> 
        </dt>
        <dd>Plot marginal coverage of an interval</dd>
      </dl><dl><dt>
          
          <code><a href="plot_peak_coverage_density.html">plot_peak_coverage_density()</a></code> 
        </dt>
        <dd>Plot the coverage density of the peaks</dd>
      </dl><dl><dt>
          
          <code><a href="plot_peak_coverage_distribution.html">plot_peak_coverage_distribution()</a></code> 
        </dt>
        <dd>Plot the coverage distribution of the peaks</dd>
      </dl><dl><dt>
          
          <code><a href="plot_peak_length_distribution.html">plot_peak_length_distribution()</a></code> 
        </dt>
        <dd>Plot the distribution of the peak length</dd>
      </dl><dl><dt>
          
          <code><a href="plot_peak_max_cov_distribution.html">plot_peak_max_cov_distribution()</a></code> 
        </dt>
        <dd>Plot the maximal per-cell coverage distribution of the peaks</dd>
      </dl><dl><dt>
          
          <code><a href="plot_region_mat.html">plot_region_mat()</a></code> 
        </dt>
        <dd>Plot a genomic region given a matrix</dd>
      </dl><dl><dt>
          
          <code><a href="plot_rna_markers.html">plot_rna_markers()</a></code> 
        </dt>
        <dd>Plot a heatmap of marker genes fold change over metacells given an McPeaks object</dd>
      </dl><dl><dt>
          
          <code><a href="plot_rna_markers_mat.html">plot_rna_markers_mat()</a></code> 
        </dt>
        <dd>Plot a heatmap of marker genes fold change over metacells</dd>
      </dl><dl><dt>
          
          <code><a href="plot_tracks_at_locus.html">plot_tracks_at_locus()</a></code> 
        </dt>
        <dd>Plot metacell tracks around locus</dd>
      </dl><dl><dt>
          
          <code><a href="project_atac_on_mc.html">project_atac_on_mc()</a></code> <code><a href="project_atac_on_mc.html">project_atac_on_mc_from_metacell1()</a></code> <code><a href="project_atac_on_mc.html">project_atac_on_mc_from_h5ad()</a></code> 
        </dt>
        <dd>Given metacells (usually from RNA data), project ATAC counts to get a McPeaks object</dd>
      </dl><dl><dt>
          
          <code><a href="rna_atac_cor_knn.html">rna_atac_cor_knn()</a></code> 
        </dt>
        <dd>Match every gene with the k ATAC peaks most correlated to it</dd>
      </dl><dl><dt>
          
          <code><a href="run_app.html">run_app()</a></code> 
        </dt>
        <dd>Run a shiny app for viewing the MCT object</dd>
      </dl><dl><dt>
          
          <code><a href="save_pheatmap.html">save_pheatmap()</a></code> 
        </dt>
        <dd>Function to save pheatmaps to disk while showing them on screen</dd>
      </dl><dl><dt>
          
          <code><a href="scc_extract.html">scc_extract()</a></code> 
        </dt>
        <dd>Extract summary statistics per cell for a set of intervals</dd>
      </dl><dl><dt>
          
          <code><a href="scc_project_on_mc.html">scc_project_on_mc()</a></code> 
        </dt>
        <dd>Given metacells (usually from RNA data), project ATAC counts to get a McCounts object</dd>
      </dl><dl><dt>
          
          <code><a href="scc_read.html">scc_read()</a></code> 
        </dt>
        <dd>Read an ScCounts object from a directory</dd>
      </dl><dl><dt>
          
          <code><a href="scc_write.html">scc_write()</a></code> <code><a href="scc_write.html">mcc_write()</a></code> 
        </dt>
        <dd>Write a counts object to a file</dd>
      </dl><dl><dt>
          
          <code><a href="set_parallel.html">set_parallel()</a></code> 
        </dt>
        <dd>Set parallel threads</dd>
      </dl><dl><dt>
          
          <code><a href="split_long_peaks.html">split_long_peaks()</a></code> 
        </dt>
        <dd>Split long peaks into smaller peaks</dd>
      </dl><dl><dt>
          
          <code><a href="subset_peak_clusters.html">subset_peak_clusters()</a></code> 
        </dt>
        <dd>Subset McPeaks by certain clusters</dd>
      </dl><dl><dt>
          
          <code><a href="subset_peaks.html">subset_peaks()</a></code> 
        </dt>
        <dd>Subset peaks of an McPeaks object</dd>
      </dl><dl><dt>
          
          <code><a href="temp_track_name.html">temp_track_name()</a></code> 
        </dt>
        <dd>Get a temporary track name</dd>
      </dl><dl><dt>
          
          <code><a href="write_sc_counts_from_bam.html">write_sc_counts_from_bam()</a></code> 
        </dt>
        <dd>Write single cell counts from an indexed bam file</dd>
      </dl><dl><dt>
          
          <code><a href="write_sc_counts_from_fragments.html">write_sc_counts_from_fragments()</a></code> 
        </dt>
        <dd>Write single cell counts from a 10x fragments file</dd>
      </dl><dl><dt>
          
          <code><a href="write_sparse_matrix_from_bam.html">write_sparse_matrix_from_bam()</a></code> 
        </dt>
        <dd>Write a sparse-matrix file from an indexed bam file</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Aviezer Lifshitz, Yonatan Shapira.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

    </footer></div>

  

  

  </body></html>

