<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function plots a set of tracks in a certain locus. Tracks can be given as a vector (tracks),
or retreived using track_regex. The intervals to be plotted can be chosen explicitly using the parameter
intervals. Alternatively, one can specify the parameter gene; by default, the gene body (all exons)
will be plotted, this can be modified using the parameters extend (how many bp to extend from each side of
the gene features) and gene_feature, where the intervals can be centered around TSSs of
the gene (gene_feature = 'tss'). 
Metacell annotations can be added using annotation_row and annotation_colors (phetamap-style annotations),
which can/should be generated automatically by generate_pheatmap_annotation.
RNA expression values of the gene of interest can be added by specifying an McATAC object, atac, to
which an RNA metacell object is attached (see the function add_mc_rna)"><title>Plot metacell tracks around locus — plot_tracks_at_locus • mcATAC</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plot metacell tracks around locus — plot_tracks_at_locus"><meta property="og:description" content="This function plots a set of tracks in a certain locus. Tracks can be given as a vector (tracks),
or retreived using track_regex. The intervals to be plotted can be chosen explicitly using the parameter
intervals. Alternatively, one can specify the parameter gene; by default, the gene body (all exons)
will be plotted, this can be modified using the parameters extend (how many bp to extend from each side of
the gene features) and gene_feature, where the intervals can be centered around TSSs of
the gene (gene_feature = 'tss'). 
Metacell annotations can be added using annotation_row and annotation_colors (phetamap-style annotations),
which can/should be generated automatically by generate_pheatmap_annotation.
RNA expression values of the gene of interest can be added by specifying an McATAC object, atac, to
which an RNA metacell object is attached (see the function add_mc_rna)"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">mcATAC</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/mcATAC.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/peaks.html">peaks</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/tanaylab/mcATAC/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Plot metacell tracks around locus</h1>
      <small class="dont-index">Source: <a href="https://github.com/tanaylab/mcATAC/blob/HEAD/R/plot.R" class="external-link"><code>R/plot.R</code></a></small>
      <div class="d-none name"><code>plot_tracks_at_locus.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function plots a set of tracks in a certain locus. Tracks can be given as a vector (<code>tracks</code>),
or retreived using <code>track_regex</code>. The intervals to be plotted can be chosen explicitly using the parameter
<code>intervals</code>. Alternatively, one can specify the parameter <code>gene</code>; by default, the gene body (all exons)
will be plotted, this can be modified using the parameters <code>extend</code> (how many bp to extend from each side of
the gene features) and <code>gene_feature</code>, where the intervals can be centered around TSSs of
the gene (gene_feature = 'tss'). <br>
Metacell annotations can be added using <code>annotation_row</code> and <code>annotation_colors</code> (phetamap-style annotations),
which can/should be generated automatically by <code>generate_pheatmap_annotation</code>.
RNA expression values of the <code>gene</code> of interest can be added by specifying an McATAC object, <code>atac</code>, to
which an RNA metacell object is attached (see the function <code>add_mc_rna</code>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">plot_tracks_at_locus</span><span class="op">(</span>
  tracks <span class="op">=</span> <span class="cn">NULL</span>,
  gene <span class="op">=</span> <span class="cn">NULL</span>,
  intervals <span class="op">=</span> <span class="cn">NULL</span>,
  iterator <span class="op">=</span> <span class="fl">50</span>,
  extend <span class="op">=</span> <span class="fl">0</span>,
  gene_feature <span class="op">=</span> <span class="st">"exon"</span>,
  track_regex <span class="op">=</span> <span class="cn">NULL</span>,
  chromosomes <span class="op">=</span> <span class="fu">gintervals.all</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"_"</span>, <span class="va">chrom</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu">pull</span><span class="op">(</span><span class="va">chrom</span><span class="op">)</span>,
  atac <span class="op">=</span> <span class="cn">NULL</span>,
  order_rows <span class="op">=</span> <span class="cn">FALSE</span>,
  row_order <span class="op">=</span> <span class="cn">NULL</span>,
  log_transform <span class="op">=</span> <span class="cn">TRUE</span>,
  gene_annot <span class="op">=</span> <span class="cn">TRUE</span>,
  name <span class="op">=</span> <span class="cn">NULL</span>,
  rna_vals <span class="op">=</span> <span class="cn">NULL</span>,
  annotation_row <span class="op">=</span> <span class="cn">NULL</span>,
  annotation_col <span class="op">=</span> <span class="cn">NULL</span>,
  annotation_colors <span class="op">=</span> <span class="cn">NULL</span>,
  silent <span class="op">=</span> <span class="cn">FALSE</span>,
  scale_bar_length <span class="op">=</span> <span class="cn">NULL</span>,
  rna_legc_eps <span class="op">=</span> <span class="fl">0.00001</span>,
  smooth_mcs <span class="op">=</span> <span class="fl">1</span>,
  smooth_bins <span class="op">=</span> <span class="fl">1</span>,
  colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"lightblue"</span>, <span class="st">"blue"</span>, <span class="st">"darkred"</span>, <span class="st">"yellow"</span><span class="op">)</span>,
  color_breaks <span class="op">=</span> <span class="cn">NULL</span>,
  use_raster <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>tracks</dt>
<dd><p>(optional) all tracks to plot</p></dd>
<dt>gene</dt>
<dd><p>(optional) which gene to plot around; if no gene is found as a whole, partial matches are searched for</p></dd>
<dt>intervals</dt>
<dd><p>(optional) what genomic interval to plot</p></dd>
<dt>iterator</dt>
<dd><p>(optional; default - 50) misha iterator, determines the resolution at which the data is extracted and plotted</p></dd>
<dt>extend</dt>
<dd><p>(optional; default - 0) how much to extend <code>intervals</code> by on each side</p></dd>
<dt>gene_feature</dt>
<dd><p>(optional; default - "exon") whether to use the exon or tss as features of the gene on which to center the plot</p></dd>
<dt>track_regex</dt>
<dd><p>(optional) regular expression for matching tracks to plot</p></dd>
<dt>chromosomes</dt>
<dd><p>(optional) which set of chromosomes to use; mainly used to filter out contigs by default</p></dd>
<dt>atac</dt>
<dd><p>(optional) McATAC object from which to extract peaks in locus, and possibly <br>
RNA expression values (if an RNA metacell object is attached)</p></dd>
<dt>order_rows</dt>
<dd><p>whether to order the rows by <code>cell_type</code> annotation</p></dd>
<dt>row_order</dt>
<dd><p>an explicit ordering of rows/tracks to use</p></dd>
<dt>log_transform</dt>
<dd><p>transfrom the data to log2(1 + x). Default: TRUE</p></dd>
<dt>gene_annot</dt>
<dd><p>(optional) whether to add gene annotations; these annotations rely on the existence of an <code>annots/refGene.txt</code> file in the genome's misha directory</p></dd>
<dt>name</dt>
<dd><p>(optional; default - 'ATAC signal') name to give the legend of the heatmap</p></dd>
<dt>rna_vals</dt>
<dd><p>(optional) RNA expression values to plot (overrides RNA expression extracted from <code>atac</code>)</p></dd>
<dt>annotation_row</dt>
<dd><p>pheatmap-format annotation</p></dd>
<dt>annotation_col</dt>
<dd><p>pheatmap-format annotation</p></dd>
<dt>annotation_colors</dt>
<dd><p>pheatmap-format annotation</p></dd>
<dt>silent</dt>
<dd><p>(optional) whether to print generated plot</p></dd>
<dt>rna_legc_eps</dt>
<dd><p>(optional) what regularization value to add to mc_rna@e_gc when calculating log</p></dd>
<dt>smooth_mcs</dt>
<dd><p>number of mc's for smoothign window (Default 1)</p></dd>
<dt>smooth_bins</dt>
<dd><p>number of genomic bins for smoothing window (default 1)</p></dd>
<dt>colors</dt>
<dd><p>(optional) vector of colors out of which the shades of the heatmap will be generated</p></dd>
<dt>color_breaks</dt>
<dd><p>(optional) defining the absolute breaks of the colors. If NULL - this will be determined as uniform distribution with clipping on the 99 percentile of the smoothed value distribution.</p></dd>
<dt>use_raster</dt>
<dd><p>Whether render the heatmap body as a raster image. It helps to reduce file size when the matrix is huge. If number of rows or columns is more than 2000, it is by default turned on. Note if <code>cell_fun</code> is set, <code>use_raster</code> is enforced to be <code>FALSE</code>.</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a ComplexHeatmap figure</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in"><span class="va">atac_sc</span> <span class="op">&lt;-</span> <span class="fu"><a href="import_from_10x.html">import_from_10x</a></span><span class="op">(</span><span class="st">"pbmc_data"</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, id <span class="op">=</span> <span class="st">"PBMC"</span>, description <span class="op">=</span> <span class="st">"PBMC from a healthy donor - granulocytes removed through cell sorting (10k)"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">cell_to_metacell_pbmc_example</span><span class="op">)</span></span>
<span class="r-in"><span class="va">atac_mc</span> <span class="op">&lt;-</span> <span class="fu"><a href="project_atac_on_mc.html">project_atac_on_mc</a></span><span class="op">(</span><span class="va">atac_sc</span>, <span class="va">cell_to_metacell_pbmc_example</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">mcmd</span><span class="op">)</span></span>
<span class="r-in"><span class="va">atac_mc</span> <span class="op">&lt;-</span> <span class="fu"><a href="add_mc_metadata.html">add_mc_metadata</a></span><span class="op">(</span><span class="va">atac_mc</span>, <span class="va">mcmd</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">rna_mc_mat</span><span class="op">)</span></span>
<span class="r-in"><span class="va">atac_mc</span> <span class="op">&lt;-</span> <span class="fu"><a href="add_mc_rna.html">add_mc_rna</a></span><span class="op">(</span><span class="va">atac_mc</span>, <span class="va">rna_mc_mat</span><span class="op">)</span></span>
<span class="r-in"><span class="va">pbmc_tracks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"unnorm"</span>, <span class="fu">gtrack.ls</span><span class="op">(</span><span class="st">"PBMC_10X.mc"</span><span class="op">)</span>, inv <span class="op">=</span> <span class="cn">TRUE</span>, v <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span class="r-in"><span class="co"># plot gene-centered intervals</span></span>
<span class="r-in"><span class="fu">plot_tracks_at_locus</span><span class="op">(</span></span>
<span class="r-in">    tracks <span class="op">=</span> <span class="va">pbmc_tracks</span>, extend <span class="op">=</span> <span class="fl">5e+4</span>,</span>
<span class="r-in">    gene <span class="op">=</span> <span class="st">"ATF3"</span>,</span>
<span class="r-in">    atac <span class="op">=</span> <span class="va">atac_mc</span>,</span>
<span class="r-in">    gene_annot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span class="r-in">    order_rows <span class="op">=</span> <span class="cn">TRUE</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># plot arbitrary intervals</span></span>
<span class="r-in"><span class="fu">plot_tracks_at_locus</span><span class="op">(</span></span>
<span class="r-in">    tracks <span class="op">=</span> <span class="va">pbmc_tracks</span>,</span>
<span class="r-in">    intervals <span class="op">=</span> <span class="fu">gintervals</span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1e+7</span>, <span class="fl">1.15e+7</span><span class="op">)</span>,</span>
<span class="r-in">    atac <span class="op">=</span> <span class="va">atac_mc</span>,</span>
<span class="r-in">    gene_annot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span class="r-in">    order_rows <span class="op">=</span> <span class="cn">TRUE</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Aviezer Lifshitz, Yonatan Shapira.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer></div>

  

  

  </body></html>

