<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='This function reads a bam file and writes a sparse-matrix file, where rows are the genomic coordinates
and columns are the cells. It does so by:
filtering the bam file by the given region (e.g. chromosom, using the "regions" paramter)
extracting the cell name using the "CB" tag added by cellranger, while excluding reads which were marked as PCR duplicates (1024)
use awk to extract the start coordinates and the tag
use unix "sort" and "uniq" to count the number of times each coordinate appears for each cell
use an R script to convert the cell names to indices
write the sparse-matrix file (matrix-market format) and zip it

Note (1): In order to use this function, "samtools" (&amp;gt;= 1.15), "awk", "sed", "sort" and "uniq" must be available at the unix command line.
Note (2): The function takes a while to run - around 13 minutes using 24 cores on the PBMC data.'><title>Write a sparse-matrix file from an indexed bam file — write_sparse_matrix_from_bam • mcATAC</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Write a sparse-matrix file from an indexed bam file — write_sparse_matrix_from_bam"><meta property="og:description" content='This function reads a bam file and writes a sparse-matrix file, where rows are the genomic coordinates
and columns are the cells. It does so by:
filtering the bam file by the given region (e.g. chromosom, using the "regions" paramter)
extracting the cell name using the "CB" tag added by cellranger, while excluding reads which were marked as PCR duplicates (1024)
use awk to extract the start coordinates and the tag
use unix "sort" and "uniq" to count the number of times each coordinate appears for each cell
use an R script to convert the cell names to indices
write the sparse-matrix file (matrix-market format) and zip it

Note (1): In order to use this function, "samtools" (&amp;gt;= 1.15), "awk", "sed", "sort" and "uniq" must be available at the unix command line.
Note (2): The function takes a while to run - around 13 minutes using 24 cores on the PBMC data.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">mcATAC</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/mcATAC.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/analysis.html">Analysis</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/tanaylab/mcATAC/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Write a sparse-matrix file from an indexed bam file</h1>
      <small class="dont-index">Source: <a href="https://github.com/tanaylab/mcATAC/blob/HEAD/R/import-counts.R" class="external-link"><code>R/import-counts.R</code></a></small>
      <div class="d-none name"><code>write_sparse_matrix_from_bam.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function reads a bam file and writes a sparse-matrix file, where rows are the genomic coordinates
and columns are the cells. It does so by:</p><ol><li><p>filtering the bam file by the given region (e.g. chromosom, using the "regions" paramter)</p></li>
<li><p>extracting the cell name using the "CB" tag added by cellranger, while excluding reads which were marked as PCR duplicates (1024)</p></li>
<li><p>use awk to extract the start coordinates and the tag</p></li>
<li><p>use unix "sort" and "uniq" to count the number of times each coordinate appears for each cell</p></li>
<li><p>use an R script to convert the cell names to indices</p></li>
<li><p>write the sparse-matrix file (matrix-market format) and zip it</p></li>
</ol><p><br>
Note (1): In order to use this function, "samtools" (&gt;= 1.15), "awk", "sed", "sort" and "uniq" must be available at the unix command line.
Note (2): The function takes a while to run - around 13 minutes using 24 cores on the PBMC data.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">write_sparse_matrix_from_bam</span><span class="op">(</span></span>
<span>  <span class="va">bam_file</span>,</span>
<span>  <span class="va">out_file</span>,</span>
<span>  <span class="va">cell_names</span>,</span>
<span>  <span class="va">region</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_mapq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  samtools_bin <span class="op">=</span> <span class="st">"samtools"</span>,</span>
<span>  samtools_opts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  num_reads <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>bam_file</dt>
<dd><p>name of the bam file</p></dd>


<dt>out_file</dt>
<dd><p>name of the output file. A ".gz" extension will be added to the file name.</p></dd>


<dt>cell_names</dt>
<dd><p>a vector with the cell names or an ATAC object</p></dd>


<dt>region</dt>
<dd><p>intervals set to filter. See samtools docs <a href="http://www.htslib.org/doc/samtools-view.html" class="external-link">http://www.htslib.org/doc/samtools-view.html</a> for details</p></dd>


<dt>genome</dt>
<dd><p>genome name (e.g. hg19). Will be inferred from the ScPeaks object if provided</p></dd>


<dt>min_mapq</dt>
<dd><p>minimal mapping quality (optional)</p></dd>


<dt>samtools_bin</dt>
<dd><p>path to samtools executable</p></dd>


<dt>samtools_opts</dt>
<dd><p>additional options for samtools (e.g. "--subsample 0.1")</p></dd>


<dt>num_reads</dt>
<dd><p>number of reads (within the <code>region</code>) to process (optional).</p></dd>


<dt>verbose</dt>
<dd><p>verbose output (optional)</p></dd>


<dt>overwrite</dt>
<dd><p>overwrite existing files (optional)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>None</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="fu">write_sparse_matrix_from_bam</span><span class="op">(</span><span class="st">"pbmc_data/possorted_bam.bam"</span>, <span class="st">"chrom1.mm"</span>, region <span class="op">=</span> <span class="fu"><a href="https://tanaylab.github.io/misha/reference/gintervals.all.html" class="external-link">gintervals.all</a></span><span class="op">(</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span>, cell_names <span class="op">=</span> <span class="va">atac_sc</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Aviezer Lifshitz, Yonatan Shapira.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

